Integrate with core Git subcommand authoring features.

tomeon opened the pull request:
Thank you for this handy tool!

This PR introduces integration with `git-sh-setup`, a library included with the
core Git distribution.  It provides conveniences for authoring shell-based Git
subcommands.

Summary of features:

1. Added help text (`git redate -h`).
2. Added CLI options sanitation (`git redate` now bails upon unrecognized options).
3. Updated to use `$GIT_EDITOR` / `git var GIT_EDITOR` for editing the re-dating specification file.  This is a breaking change, though one that's likely to align `git redate` with the user's already-expressed preference (e.g. whatever they've defined as `core.editor`) rather than prompting them to express a preference interactively.
4. Added support for specifying `git redate` preferences in the Git client configuration file.
5. Added support for specifying arbitrary commit ranges.

Miscellaneous other changes:

1. Streamlined some code (e.g. removing repeated checks of the value of `$ALL`, `$DEBUG`, etc., instead using conditionally-defined functions).
2. Prevent trying to `shift` any arguments when `$@` is already empty.
3. Added a dry-run mode, which was helpful for testing the other changes and perhaps useful for end users before re-dating For Real.
4. Fixed some `shellcheck` linting issues.
5. Formatted the code with `shfmt`.
6. Relocated the temporary file used for specifying date edits to the `.git` directory.

Thanks for your consideration!

tomeon committed:
feat: use `/bin/sh` interpreter

rather than non-portable `/bin/bash`

tomeon committed:
fix: replace Bash-specific code

with POSIX-compatible code.

tomeon committed:
feat: integrate with git's core shell library

for features like CLI options processing and "am I in a git repo?"
sanity-checks.

tomeon committed:
feat: put tmpfile in `.git`

tomeon committed:
fix: do not attempt to shift > ARGC arguments

That is, remove redundant `shift` calls in the `--debug` and `--all`
cases.

tomeon committed:
chore: fix shellcheck linting errors

tomeon committed:
chore: simplify debugging code

by conditionally defining a `debug` function based on whether the
`-d/--debug` option was provided.  If debugging is not enabled, the
`debug` function is a NOP; otherwise, it echoes its arguments to stderr.

tomeon committed:
feat: add dry-run mode

tomeon committed:
chore: send diagnostics to stderr

BREAKING CHANGE: diagnostic output no longer goes to stdout.

tomeon committed:
BREAKING CHANGE: use `$GIT_EDITOR` for date edits

rather than custom editor selection code.

tomeon committed:
feat: support `redate.<param>` git config settings

BREAKING CHANGE: introduces type-checking of `COMMITS`, `DEBUG`, and
friends.

tomeon committed:
chore: remove redundant `git log`

that provides a default value for the `COMMITS` variable.  This variable
is already guaranteed to be set to a positive integer (or else `git
redate` will have exited).

tomeon committed:
chore: format code with shfmt

tomeon committed:
feat: remove repeated `--all` option checks

within the body of the loop that applies the redate logic.  Instead,
conditionally define an `apply_redate` function depending on whether we
saw `--all`, then call this function from within the loop.

tomeon committed:
fix: add message for last chunk to be collected

tomeon committed:
feat: permit specifying commit ranges

rather than numbers of commits starting from HEAD (or all commits).

tomeon committed:
chore: remove extraneous whitespace

tomeon committed:
feat: show examples of usage modes

in help text

tomeon committed:
doc: describe new options and configuration params

iitzrohan commented:
@tomeon Thanks for updating the tool.Can you add two more features.

1. Don't change the hash when redating.
2. Make it possible to sign commits with redating.

tomeon committed:
fix: omit previously-filtered commits

when generating the interactive commit selection list.

tomeon committed:
feat: support signing rewritten commits

by passing the `-S`/`--gpg-sign` flag through to `commit-tree`.

tomeon committed:
fix: normalize Git configuration settings

by loading them via `git -c <name>=<value> --get <name>`.

tomeon commented:
Hello @iitzrohan -- thanks for your feedback/requests.

> Don't change the hash when redating.

This is not possible.

> Make it possible to sign commits with redating.

Done.

iitzrohan commented:
The latest changes doesn't opens the editor configured through gitconfig. Instead it uses the default editor nano.

iitzrohan commented:
@tomeon  Also getting this error while redate.

```
fatal: ambiguous argument 'HEAD~1..HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Git redate failed. Please make sure you run this on a clean working directory.

```

tomeon committed:
fix: exit if no commits matched selection criteria

BREAKING CHANGE: `git redate` exits with nonzero status in case no
commits were identified.

tomeon commented:
@iitzrohan -- 

> The latest changes doesn't opens the editor configured through gitconfig. Instead it uses the default editor nano.

I am unable to reproduce this; for me, `git redate` respects `core.editor`.

> `fatal: ambiguous argument [...]`

Thanks; should be fixed now.  I've updated `git-redate` to check the number of commits reachable from `HEAD` and choose the lower of (a) that number and (b) the number provided with `git redate -c <count>`.  I've also updated `git-redate` to exit with nonzero status in case the provided commit range is invalid, or valid but causes `git rev-list` to emit no commits.

plasticrake commented on git-redate:
```diff
@@ -1,209 +1,357 @@
+
+# Allow running this command from a subdirectory of the working tree
+# shellcheck disable=SC1091
+SUBDIRECTORY_OK=yes . git-sh-setup
```
In order for this to work on my system I had to change this line:
```sh
SUBDIRECTORY_OK=yes . "$(git --exec-path)/git-sh-setup"
```
Taken from: https://git-scm.com/docs/git-sh-setup

tomeon replied:
Thanks.  I assume you are invoking this via `git-redate` directly rather than [installing `git-redate` into a `$PATH` entry](https://github.com/PotatoLabs/git-redate#installation)?  In any case, I guess there's no harm in fully-qualifiying the `git-sh-setup` path rather than relying on a `$PATH` search that depends on the script being run as `git redate`.

tomeon replied:
This is done.

tomeon commented:
Ping @AaronVasquez @EdmundMai -- this changeset, besides being (IMHO :wink:) worthwhile on its own merits, also likely fixes a number of issues.  For instance:

1. #6, #21, #39 
2. #38 
3.  Probably #10, #31, #38

It also addresses or obviates several merge requests:

1. #29
2. #30
3. #33
4. #41 

Is there anything I can do to help get this over the finish line?  Thanks!
